pennylane/pulse/transmon.py

```diff
@@ -0,0 +1,193 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This module contains the classes/functions specific for simulation of superconducting transmon hardware systems"""
+from dataclasses import dataclass
+from typing import Callable, List, Union
+
+import pennylane as qml
+from pennylane.pulse import HardwareHamiltonian
+from pennylane.typing import TensorLike
+
+
+# pylint: disable=unused-argument
+def a(wire, d=2):
+    """creation operator"""
+    return qml.s_prod(0.5, qml.PauliX(wire)) + qml.s_prod(0.5j, qml.PauliY(wire))
+
+
+def ad(wire, d=2):
+    """annihilation operator"""
+    return qml.s_prod(0.5, qml.PauliX(wire)) + qml.s_prod(-0.5j, qml.PauliY(wire))
+
+
+def transmon_interaction(
+    connections: list, omega: Union[float, list], g: Union[float, list], delta=None, wires=None, d=2
+):
+    r"""Returns a :class:`ParametrizedHamiltonian` representing the cQED Hamiltonian of a superconducting transmon system.
+
+    The Hamiltonian is given by
+
+    .. math::
+
+        H = \sum_q \omega_q a^\dagger_q a_q + \sum_{(i, j) \in \mathcal{C}} g_{ij} \left(a^\dagger_i a_j + a_j^\dagger a_i \right)
+        + \sum_q \delta_q a^\dagger_q a^\dagger_q a_q a_q
+
+    where :math:`[a^\dagger_p, a_q] = i \delta_{pq}` are bosonic creation and annihilation operators.
+    The first term describes the dressed qubit frequencies :math:`\omega_q`, the second term their coupling :math:`g_{ij}` and the last the anharmonicity :math:`\delta_q`,
+    which all can vary for different qubits. In practice, the bosonic operators are restricted to a finite dimension of the local Hilbert space (default ``d=2`` corresponds to qubits).
+    In that case, the anharmonicity is set to :math:`delta=0` and ignored.
+
+    .. note:: Currently only supporting ``d=2`` with qudit support planned in the future.
+
+    TODO: resource / source for good values.
+
+    .. seealso::
+
+        :func:`~.drive`
+
+    Args:
+        connections (list[tuple(int)]): List of connections ``(i, j)`` between qubits i and j.
+        omega (Union[float, list[float]]): List of dressed qubit frequencies in GHz.
+            When passing a single float all qubits are assumed to have that same frequency.
+        g (Union[float, list[float]]): List of coupling strengths in GHz. Needs to match the length of ``connections``.
+            When passing a single float all connections are assumed to have the same strength.
+        delta (Union[float, list[float]]): List of anharmonicities in GHz. Ignored when ``d=2``.
+            When passing a single float all qubits are assumed to have that same anharmonicity.
+        wires (list): Optional, defaults to the unique wires in ``connections`` when set to ``None``. When passing explicit ``wires``,
+            needs to at least contain all unique wires in ``connections`` and can be used to initiate additional, unconnected qubits.
```

Comment:
I shifted to a focus of `wires` <> `omega`, and the connection terms are kind of independent. This may lead to situations where the passed `wires` are not equal to the op`s wires, which is mentioned in the strings and (kind of?) makes sense:
```python
>>> transmon_drive(omega=1, connections=[[0,5], [7, 8]], g=[0.5, 0.5], wires=range(5)).wires
<Wires = [0, 1, 2, 3, 4, 5, 7, 8]>
```
What do you think?

Reply:
Currently we get things like this:

```
>>> op = transmon_interaction(omega=[6.7, 6.8, 6.9], connections=[[0,5], [7, 8]], g=[0.5, 0.5])
>>> op.wires
<Wires = [0, 1, 2, 5, 7, 8]>
>>> op.settings
TransmonSettings(connections=[[0, 5], [7, 8]], omega=[6.7, 6.8, 6.9], g=[0.5, 0.5], anharmonicity=[0.0, 0.0, 0.0])
```

Does this system have 3 qubits (it has 3 resonant frequencies) or does it have 6 qubits (it has 6 wires)? And which qubits are coupled? Do qubits 7 and 8 have a resonant frequency? Etc. 

I think treating `len(omega)` as the only source of truth for the number of qubits in the system, with wires only being used to rename them, and everything else compared to the resulting wires, is unambiguous and closely aligned with the physical system, and that would be my go-to solution. Other configurations could also be reasonable. But currently I think the behaviour is still unintuitive and a bit buggy.

Reply:
I think this discussion is the same / related to https://github.com/PennyLaneAI/pennylane/pull/3936#discussion_r1162544419